storage/innobase/log/log0online.cc

```diff
@@ -1132,23 +1268,70 @@ static bool log_online_follow_log_seg(
-      log_bmp_sys->read_buf.get_current_lsn()) {
-    // Seeking to the new read data to start parsing from, next block must be
-    // fully or partially new data to parse
+  ut_ad(log_bmp_sys->read_buf.get_current_lsn() + OS_FILE_LOG_BLOCK_SIZE >
+        log_bmp_sys->start_lsn);
+
+  ulint skip_len{0};
+
+  // If we get here it means all blocks before one containing
+  // log_bmp_sys->start_lsn are already skipped. Now check if we know where
+  // to start parsing the log.
+  if (log_bmp_sys->parse_buf.get_parse_start_lsn() == 0) {
+    // This is the beginning of the iteration for a new checkpoint. At this
+    // point log_bmp_sys->start_lsn contains the previous checkpoint LSN.
+    // As it may point into a middle of a log record we need to find first
+    // record group which starts at or after checkpoint LSN and ignore
+    // everything before this point.
+    auto first_rec_group_offset{0};
+
+    // Looking for the first complete record to start parsing at.
+    while (log_bmp_sys->read_buf.is_data_available() &&
+           (first_rec_group_offset = log_block_get_first_rec_group(
+                log_bmp_sys->read_buf.ccurrent())) == 0) {
+      log_bmp_sys->read_buf.advance();
+    }
+
+    if (!log_bmp_sys->read_buf.is_data_available()) {
+      // Go to the next log segment.
+      return true;
+    }
+
+    ut_ad(first_rec_group_offset > 0);
+
+    const auto parse_start_lsn =
+        log_bmp_sys->read_buf.get_current_lsn() + first_rec_group_offset;
+
+    if (parse_start_lsn >= log_bmp_sys->start_lsn) {
+      log_bmp_sys->parse_buf.set_parse_start_lsn(parse_start_lsn);
+    }
+
+    // Even if parse_start_lsn points before log_bmp_sys->start_lsn we still
+    // need to skip incomplete record at the beginning of a block.
+    skip_len = static_cast<ulint>(first_rec_group_offset);
+    // In case first record group in this block starts before
+    // log_bmp_sys->start_lsn, parse start LSN will be found later while parsing
+    // log records.
+  } else if (log_bmp_sys->has_parse_data_to() >
+             log_bmp_sys->read_buf.get_current_lsn()) {
+    // This is just another segment processing. Skip data which is already in
+    // parse buffer from previous iteration.
     ut_ad(log_bmp_sys->read_buf.get_current_lsn() <
           log_bmp_sys->has_parse_data_to());
     ut_ad(log_bmp_sys->read_buf.get_current_lsn() + OS_FILE_LOG_BLOCK_SIZE >
           log_bmp_sys->has_parse_data_to());
 
-    /* How many bytes of log data should we skip in the current log block. */
-    const auto skip_already_in_parse_buf_len =
-        static_cast<ulint>(log_bmp_sys->has_parse_data_to() -
-                           log_bmp_sys->read_buf.get_current_lsn());
-    log_online_parse_redo_log_block(log_bmp_sys->read_buf,
-                                    skip_already_in_parse_buf_len);
+    skip_len = static_cast<ulint>(log_bmp_sys->has_parse_data_to() -
+                                  log_bmp_sys->read_buf.get_current_lsn());
```

Comment:
where is the code to skip the empty blocks on redo log resize? is this the one?

Reply:
Empty blocks are skipped in log_online_parse_redo_log_block(). And here it skips already parsed data.